### DISCUSSION TITLE
How do you write your general fetcher functions/wrappers/abstractions and error handling?

### DISCUSSION THREAD

**Discussion by bennettdams** [TanStack/query#4875] [Q&A] [answered]
Hey!

the title says it already: How do you write your fetcher functions/wrappers and how do you do error handling?
I don't think it is reasonable to always write ad-hoc fetcher functions for each `useQuery`'s `queryFn`, so I guess most people write reusable functions. I would like to see different ways of implementing the functions that are then used as `queryFn` in TanStack Query and your reasoning for the decisions. 

### General questions

- How do you usually return data from your fetcher function? Is it the data directly or rather the response itself, so the caller has access to the HTTP status etc.?
- How do you handle the unhappy path? Reject the promise? Throw an error?
- Do you do error handling in the base fetcher (`http` in my example) or in the specific fetcher (e. g. `fetchPost` in my example)? How do you deal with this on the server like Next.js (e.g. `try..catch` everywhere)? Should we only use error boundaries from now on to prepare for `Suspense`'s integration?

### Example implementation

Here's an example of how I usually do this (when I can't use tRPC 😅):

I have a generic `http` function used for fetching. I also have `get`, `put`, `post` and `delete` functions that are using `http` inside with the appropiate configuration. 
Then, for every data entity, I have a specific fetcher function (`fetchPost`, `fetchUser`, ...) that uses one of the four functions inside.

The most important one is the `http` function though:
It expects a `Request` which comes with an URL, HTTP options, etc.
It takes care of setting the HTTP header, executes the `fetch` and then does different things based on the response. It will e.g. reject the promise for specific HTTP status codes or also grabs the data from the response body.

Usually my `http` function returns a custom type (something like `HttpResponse`), which extends the standard `Response` and adds another field `result` where I put the data when everything went well.
PS: Recently I thought about skipping this step and only return the data directly. Usually TanStack Query catches a rejected Promise for me, but if the function is used somewhere else (e. g. the server to prefill the cache for hydration), this is no option anymore. But returning the data directly will make it easier for TanStack Query to infer a truthy value I think, because right now the responded data in the `result` field is always nullable (so it is of type `T | null`). If we can savely assume that a succesful execution of the `http` function always returns the truthy data, TanStack Query can narrow the type I think. Maybe I need to have a conditional type that makes sure that `T` is not `null` if we checked the response data and everything went well....

---

**Comment by louis-young** [comment, marked as answer]
Hi 👋 

This is quite a broad set of questions to which the answers often depend on your specific requirements and setup but I'll attempt to answer them as best I can. I don't think there are any objective answers to these questions; opinions and preferences vary, so take my answers with a pinch of salt.

In general, HTTP verb and API service/client abstractions are common. 

1. I usually return a promise that resolves to data. I don't typically care for the state of the underlying request in the happy path.
2. Either option is valid as far as I'm aware. I typically reject the promise. 
3. There are three questions here:
   1. That depends on what type of error handling you're describing. I think it's sensible to handle network-related errors in the HTTP abstraction.
   2. I don't know what you mean by "on the server" unfortunately. I'm not sure if you're referring to a backend API you're interfacing with or TanStack Query in a server context, or something else entirely.
   3. I don't think so, no. Suspense is still experimental so I haven't been paying _too_ much attention to it as the API and semantics are subject to change.

I imagine the official TanStack Query course will somewhat authoritatively cover topics like this to some degree if you're interested: https://ui.dev/react-query?from=tanstack

Hopefully this helps 🙂
